core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v2.go

```diff
@@ -0,0 +1,324 @@
+	Enqueue(id *big.Int, logs ...logpoller.Log) (added int, dropped int)
+	// Dequeue pulls logs from the buffer that are within the given block window,
+	// with a maximum number of logs per upkeep and a total maximum number of logs to return.
+	// It also accepts a function to select upkeeps.
+	// Returns logs (associated to upkeeps) and the number of remaining
+	// logs in that window for the involved upkeeps.
+	Dequeue(block int64, blockRate, upkeepLimit, maxResults int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int)
+	// Size returns the total number of logs in the buffer.
+	Size() int
+	// SetConfig sets the buffer size and the maximum number of logs to keep for each upkeep.
+	SetConfig(lookback, maxUpkeepLogs int)
+}
+
+func DefaultUpkeepSelector(id *big.Int) bool {
+	return true
+}
+
+type logBuffer struct {
+	lggr logger.Logger
+	// max number of logs to keep in the buffer for each upkeep per block
+	maxUpkeepLogs *atomic.Int32
+	// number of blocks to keep in the buffer
+	bufferSize *atomic.Int32
+	// last block number seen by the buffer
+	lastBlockSeen *atomic.Int64
+	// map of upkeep id to its buffer
+	upkeepBuffers map[string]*upkeepLogBuffer
+	lock          sync.RWMutex
+}
+
+func NewLogBuffer(lggr logger.Logger, size, upkeepLogLimit int) LogBuffer {
+	s := &atomic.Int32{}
+	s.Add(int32(size))
+	l := &atomic.Int32{}
+	l.Add(int32(upkeepLogLimit))
+	return &logBuffer{
+		lggr:          lggr.Named("KeepersRegistry.LogEventBufferV2"),
+		maxUpkeepLogs: l,
+		bufferSize:    s,
+		lastBlockSeen: &atomic.Int64{},
+		upkeepBuffers: make(map[string]*upkeepLogBuffer),
+	}
+}
+
+func (b *logBuffer) SetConfig(lookback, logLimitHigh int) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.bufferSize.Store(int32(lookback))
+	b.maxUpkeepLogs.Store(int32(logLimitHigh))
+
+	for _, ub := range b.upkeepBuffers {
+		ub.setConfig(logLimitHigh)
+	}
+}
+
+func (b *logBuffer) Size() int {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	size := 0
+	for _, ub := range b.upkeepBuffers {
+		size += ub.size()
+	}
+
+	return size
+}
+
+func (b *logBuffer) Enqueue(uid *big.Int, logs ...logpoller.Log) (int, int) {
+	buf, ok := b.getUpkeepBuffer(uid)
+	if !ok || buf == nil {
+		buf = newUpkeepLogBuffer(b.lggr, uid, int(b.maxUpkeepLogs.Load()*b.bufferSize.Load()))
+		b.setUpkeepBuffer(uid, buf)
+	}
+	lastBlockSeen := latestBlockNumber(logs...)
+	if b.lastBlockSeen.Load() < lastBlockSeen {
+		b.lastBlockSeen.Store(lastBlockSeen)
+	}
+	offsetBlock := b.lastBlockSeen.Load() - int64(b.bufferSize.Load())
+	if offsetBlock <= 0 {
+		offsetBlock = 1
+	}
+	added, dropped := buf.enqueue(offsetBlock, logs...)
+
+	return added, dropped
+}
+
+func (b *logBuffer) Dequeue(block int64, blockRate, upkeepLimit, maxResults int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	start, end := BlockWindow(block, blockRate)
```

Comment:
I think the `BlockWindow` function could probably live on the logBuffer type?

Reply:
Moved to buffer_v1.go `getBlockWindow`, kept it standalone as we might want to access it from multiple structs